src/meta/src/stream/stream_graph/schedule.rs

```diff
@@ -0,0 +1,243 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(clippy::explicit_iter_loop, reason = "crepe")]
+
+use std::collections::{BTreeMap, HashMap, LinkedList};
+
+use enum_as_inner::EnumAsInner;
+use itertools::Itertools;
+use risingwave_common::bail;
+use risingwave_common::hash::ParallelUnitMapping;
+use risingwave_pb::common::ParallelUnit;
+use risingwave_pb::stream_plan::DispatcherType::{self, *};
+
+use super::{CompleteStreamFragmentGraph, GlobalFragmentId as Id};
+use crate::MetaResult;
+
+type HashMappingId = usize;
+
+/// The internal distribution structure for processing in the scheduler.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum DistId {
```

Comment:
The `crepe` requires every struct to be `Copy`, so we need to use a index-map for this and use the index to represent the mapping itself.